Invalid syntax and crash on unit struct with where-clause

Against the current master branch, 4e2f7419176f9511c8f11fbd1378a3d455c09759:

```console
$ echo "mod m { struct S where A: B; }" | cargo run --bin rustfmt

mod m {
    struct S
    mod m { struct S
    where
        A: B;
}
```

The output is not valid Rust syntax.

If the input is in two files, rustfmt crashes. I am hitting this in https://github.com/dtolnay/syn/pull/536.

#### src/lib.rs

```rust
mod m;
```

#### src/m.rs

```rust
struct A where B: C;
```

```console
$ RUST_BACKTRACE=1 cargo run --bin rustfmt -- ../path/to/src/lib.rs 

thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', libcore/option.rs:355:21
stack backtrace:
   9: <core::option::Option<T>>::unwrap
  10: rustfmt_nightly::rewrite::RewriteContext::snippet
             at src/rewrite.rs:56
  11: rustfmt_nightly::comment::rewrite_missing_comment
             at src/comment.rs:823
  12: rustfmt_nightly::items::rewrite_comments_before_after_where
             at src/items.rs:2731
  13: rustfmt_nightly::items::rewrite_where_clause_rfc_style
             at src/items.rs:2528
  14: rustfmt_nightly::items::rewrite_where_clause
             at src/items.rs:2630
  15: rustfmt_nightly::items::format_generics
             at src/items.rs:2786
  16: rustfmt_nightly::items::format_unit_struct
             at src/items.rs:1183
  17: rustfmt_nightly::items::format_struct
             at src/items.rs:991
  18: rustfmt_nightly::items::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::visit_struct
             at src/items.rs:438
  19: rustfmt_nightly::visitor::FmtVisitor::visit_item
             at src/visitor.rs:399
  20: rustfmt_nightly::reorder::<impl rustfmt_nightly::visitor::FmtVisitor<'a>>::visit_items_with_reordering
             at src/reorder.rs:281
  21: rustfmt_nightly::visitor::FmtVisitor::walk_mod_items
             at src/visitor.rs:697
  22: rustfmt_nightly::visitor::FmtVisitor::format_separate_mod
             at src/visitor.rs:781
  23: <rustfmt_nightly::formatting::FormatContext<'a, T>>::format_file
             at ./src/formatting.rs:150
  24: rustfmt_nightly::formatting::format_project
             at ./src/formatting.rs:92
  25: rustfmt_nightly::formatting::<impl rustfmt_nightly::Session<'b, T>>::format_input_inner::{{closure}}
             at ./src/formatting.rs:47
  26: <scoped_tls::ScopedKey<T>>::set
             at scoped-tls-0.1.2/src/lib.rs:155
  27: syntax::with_globals::{{closure}}
             at rustc-ap-syntax-290.0.0/lib.rs:123
  28: <scoped_tls::ScopedKey<T>>::set
             at scoped-tls-0.1.2/src/lib.rs:155
  29: syntax::with_globals
             at rustc-ap-syntax-290.0.0/lib.rs:122
  30: rustfmt_nightly::formatting::<impl rustfmt_nightly::Session<'b, T>>::format_input_inner
             at ./src/formatting.rs:31
  31: <rustfmt_nightly::Session<'b, T>>::format
             at ./src/lib.rs:531
  32: rustfmt::format_and_emit_report
             at src/bin/main.rs:300
  33: rustfmt::format::{{closure}}
             at src/bin/main.rs:272
  34: <rustfmt_nightly::Session<'b, T>>::override_config
             at ./src/lib.rs:539
  35: rustfmt::format
             at src/bin/main.rs:271
  36: rustfmt::execute
             at src/bin/main.rs:199
  37: rustfmt::main
             at src/bin/main.rs:36
```

I was not aware! What is the use ? I'd appreciate any pointers ;o)
I tried this in the playground but it fails with `error[E0412]: cannot find type `B` in this scope`
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fa1205ba785feec0af7f6d8828b2895b

--- a/src/items.rs
+++ b/src/items.rs
@@ -464,7 +464,8 @@ impl<'a> FmtVisitor<'a> {
                 BracePos::Auto
             },
             self.block_indent,
-            mk_sp(span.lo(), body_start),
+            // make a span that starts right after `enum Foo`
+            mk_sp(ident.span.hi(), body_start),
             last_line_width(&enum_header),
         )
         .unwrap();
@@ -1186,7 +1187,8 @@ fn format_unit_struct(context: &RewriteContext, p: &StructParts, offset: Indent)
             context.config.brace_style(),
             BracePos::None,
             offset,
-            mk_sp(generics.span.lo(), hi),
+            // make a span that starts right after `struct Foo`
+            mk_sp(p.ident.span.hi(), hi),
             last_line_width(&header_str),
         )?
     } else {
@@ -1208,7 +1210,7 @@ pub fn format_struct_struct(
     let header_str = struct_parts.format_header(context);
     result.push_str(&header_str);
 
-    let header_hi = span.lo() + BytePos(header_str.len() as u32);
+    let header_hi = struct_parts.ident.span.hi();
     let body_lo = context.snippet_provider.span_after(span, "{");
 
     let generics_str = match struct_parts.generics {
@@ -1222,6 +1224,7 @@ pub fn format_struct_struct(
                 BracePos::Auto
             },
             offset,
+            // make a span that starts right after `struct Foo`
             mk_sp(header_hi, body_lo),
             last_line_width(&result),
         )?,

--- a/tests/source/issue-3194.rs
+++ b/tests/source/issue-3194.rs
@@ -1,0 +1,13 @@
+mod m { struct S where A: B; }
+
+mod n { struct Foo where A: B { foo: usize } }
+
+mod o { enum Bar where A: B { Bar } }
+
+mod with_comments {
+  mod m { struct S /* before where */ where A: B; /* after where */ }
+  
+  mod n { struct Foo /* before where */ where A: B /* after where */ { foo: usize } }
+  
+  mod o { enum Bar /* before where */ where A: B /* after where */ { Bar } }
+}

--- a/tests/target/issue-3194.rs
+++ b/tests/target/issue-3194.rs
@@ -1,0 +1,52 @@
+mod m {
+    struct S
+    where
+        A: B;
+}
+
+mod n {
+    struct Foo
+    where
+        A: B,
+    {
+        foo: usize,
+    }
+}
+
+mod o {
+    enum Bar
+    where
+        A: B,
+    {
+        Bar,
+    }
+}
+
+mod with_comments {
+    mod m {
+        struct S
+        /* before where */
+        where
+            A: B; /* after where */
+    }
+
+    mod n {
+        struct Foo
+        /* before where */
+        where
+            A: B, /* after where */
+        {
+            foo: usize,
+        }
+    }
+
+    mod o {
+        enum Bar
+        /* before where */
+        where
+            A: B, /* after where */
+        {
+            Bar,
+        }
+    }
+}

